### PR TITLE
Add decorator for performing broadcasting inside translation rules

### PR DESCRIPTION
**Pull request by notEvil**
https://github.com/google/jax/issues/2163

If you accept this approach I will adjust the rest accordingly.

What are the code formatting rules? (did you consider using black or yapf?)

### CODE REVIEW

**Comment by notEvil** [comment]
@googlebot I signed it!

**Comment by shoyer** [comment]
This logic looks really similar to to `jax.numpy.vectorize`. Would it be possible to make use of that instead, or are we stuck writing things in terms of XLA HLO?

**Comment by notEvil** [comment]
I can't answer that confidently, maybe @mattjj?

**Comment by shoyer** [comment]
I think this need something _like_ the `vectorize` logic, but vectorize itself won't work because the `vmap` rule for the `igamma` primitive is broken.

**Comment by notEvil** [comment]
As discussed in https://github.com/google/jax/issues/2163, `igamma`, unlike most primitives, seems to require explicit broadcasting (.. or most primitives use TFs broadcasting, which is not available for `igamma`). That's why its implemented in the translation rule, which in turn fixes `vmap` for it (and a few others).

File: jax/lax/lax.py
Comment (hawkinsp) [inline jax/lax/lax.py:1606]: I think I'd omit the comments everywhere here, because they don't really add a whole lot that you can't also quickly see from the code. Comments that basically just restate what the code does don't help with understanding, whereas usually I want to know from a comment either "what" at a higher level or "why".

File: jax/lax/lax.py
Comment (hawkinsp) [inline jax/lax/lax.py:1604]: I'd add a brief function-level comment explaining what this function as a whole does and why it does it.

File: jax/lax/lax.py
Comment (hawkinsp) [inline jax/lax/lax.py:1608]: Primitives rules use keyword arguments for metadata, and they should never be XLA values. So you should ignore kwargs completely for the purpose of broadcasting and pass them through unchanged to the underlying translation rule.

File: jax/lax/lax.py
Comment (hawkinsp) [inline jax/lax/lax.py:184]: I think there are a few other primitives that need a similar change, e.g. `regularized_incomplete_beta`?

Could we fix them too?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks for the PR! Sorry for the slow review, we're swamped... 

Would you please also add a test for the original issue to `lax_test.py` to make sure it does not regress?

We don't have hard and fast style rules, but I suggest formatting to an 80-column width.

**Comment by notEvil** [comment]
Regarding the test, there is already one that should cover it (see https://github.com/google/jax/issues/2163#issuecomment-601266652 and later). I'll change the rest as requested

**Comment by notEvil** [comment]
I looked at the list of lax ops in `lax_test.py` and wrote a small script to check if they are affected. Only `igamma`, `igammac` and `betainc` were positive and they are all covered now. And as mentioned in the original issue, the tests are there, just need fixing (cover scalar to array broadcasting). You are free to change anything you deem necessary before merge!

**Review by hawkinsp — APPROVED** [review]
I updated this PR to apply against head and also applied the decorator to the newly added `igamma` gradient. I also made a few minor cleanups.

**Comment by notEvil** [comment]
Oh, I didn't know you can and would squash, change and force push to my fork. I would have preferred to stay author of the commits (squash myself if necessary), but if thats how you prefer it then I will obey.

**Comment by hawkinsp** [comment]
GitHub allows maintainers to push to forks that are used for PRs (as a way of collaborating on a PR). If you prefer not to allow this, you can opt out when sending a PR. But I find it's often quicker to allow collaboration, as here where I was trying to fix a merge conflict that stopped me from merging your PR. It became a little difficult to manage the merge, so I split your change out into a single commit and rebased it on top of master.

I realized I accidentally overwrote your committer details when doing so; thank for pointing that out. I fixed it.

I also fixed a new bug revealed by the merge. `igamma` now has gradients, and since the primitive itself now broadcasts internally, we must teach its JVP rule how to broadcast.

**Comment by hawkinsp** [comment]
Thanks for the PR!

**Comment by notEvil** [comment]
All good, thanks for caring :)

**Pull request merged**
